### DISCUSSION TITLE
ES6 module exports and hooks

### DISCUSSION THREAD

**Discussion by DeanKamali** [directus/directus#10656] [Archive]
The [example in docs](https://docs.directus.io/extensions/hooks/#extension-entrypoint) supports es6 module exports,  

Using the same example `extensions/hooks/test/index.js`: 
```
export default ({ filter, action }) => {
	filter('items.create', () => {
		console.log('Creating Item!');
	});

	action('items.create', () => {
		console.log('Item created!');
	});
};
```
throws this error on startup. 
```
07:16:12 ⚠️  Couldn't register hook "test"
07:16:12 ⚠️  Unexpected token 'export'
```

However, it works when using common js exports 

I have also noticed, that `init` event type never seems to receive context object as a second parameter as [per docs here](https://docs.directus.io/extensions/hooks/#register-function)
```
 init('app.after', async (app, context) => {
            console.log('started ....', context)
});
```

I get `started ...., undefined)`

**Comment by azrikahar** [comment]
> The example in docs supports es6 module exports

I believe the docs showcases examples in es6 module since it assumes you will use the extensions-sdk to build your extensions: https://docs.directus.io/extensions/creating-extensions/, where `hooks` will be output as commonjs.

In case you're curious, you can see how the extensions decides whether to output as esm or cjs here: https://github.com/directus/directus/blob/3c82d5f7eab76becb97ff02c5fa59339bcb9bfa2/packages/extensions-sdk/src/cli/commands/build.ts#L184-L198

and what are app vs api extension types: https://github.com/directus/directus/blob/3c82d5f7eab76becb97ff02c5fa59339bcb9bfa2/packages/shared/src/constants/extensions.ts#L4-L5

> I have also noticed, that init event type never seems to receive context object as a second parameter 

That's correct 👍 The docs for Init event does specify the callback function to only contain **one** parameter, which is the event-specific meta object: https://docs.directus.io/extensions/hooks/#init. It's unlike Filter or Action event that mentions the context object as the second parameter of the callback function.

**Comment by sharkfin009** [comment]
OK so this means API extensions have to be common js ? If I may ask why cant everything be in ES module syntax?

**Reply by sharkfin009** [reply]
And if the extensions CLI outputs CJS, why are docs showing ES syntax? So confused over here ;)

**Reply by rijkvanzanten** [reply]
It's complicated.. The Node.js CJS -> ESM transition has been very tricky, so it took us up until TS and Tsx support to come through to really be able to flip the switch in https://github.com/directus/directus/pull/18014 (released a week ago). Since then, everything is technically ESM by default 👍🏻 Before that, we had the extensions-sdk bundle/transpile esm/typescript down to CJS for compat with the API. The examples/templates etc were in ESM, as it seems to be the leading standard for examples 🙂

**Reply by sharkfin009** [reply]
Thanks @rijkvanzanten ! I have been looking into the ESM/CJS challenge more and I now get that its very hard to deal with.

**Reply by iddar** [reply]
I fix warning message changing the extension from `.js` to `.cjs`
